Dockerfile

```diff
@@ -0,0 +1,44 @@
+FROM ubuntu:16.04
+
+ENV DEBIAN_FRONTEND=noninteractive
+
+# Install packages needed for the build scripts
+RUN apt-get update \
+    && apt-get install -y sudo software-properties-common git wget clang-format-6.0 colordiff \
+    && apt-add-repository -y ppa:rael-gc/rvm \
+    && apt-get update \
+    && apt-get install -y rvm \
+    && apt-get autoremove --purge
+
+# Add puppet and build scripts to the image
+RUN mkdir -p /NUbots-build/toolchain/
+ADD ./puppet /NUbots-build/puppet/
+ADD ./.travis /NUbots-build/.travis/
+
+# Install the NUbots toolchain
+RUN TRAVIS_BUILD_DIR=/NUbots-build /NUbots-build/.travis/install_toolchain.sh
+
+# Remove extraneous files from the toolchain install
+RUN rm -rf /NUbots-build \
+    && apt-get autoremove --purge
+
+# Create the directory that will be used for checkouts and builds
+RUN mkdir -p /code/NUbots
+
+# Create the nubots user and group with id 1000. This will be used to map to
+# the user with id 1000 on the host (typically the first non-root user)
+# Used to fix permission issues
+RUN groupadd -g 1000 nubots \
```

Comment:
This seems suspicious... Do other examples do this?

Reply:
Yeah, from what I could find, this is normal [[1]](https://medium.com/@mccode/processes-in-containers-should-not-run-as-root-2feae3f0df3b) [[2]](https://jtreminio.com/blog/running-docker-containers-as-current-host-user/#ok-so-what-actually-works) [[3]](https://www.reddit.com/r/docker/comments/cihjtf/how_do_i_make_sure_that_files_created_by/). The problem it's trying to solve:

- The docker daemon that runs builds, containers, etc runs as root (there's no way around this, as far as I can tell)
- Root in the container is the same as root on the host -- with reduced privileges -- (since the kernel is shared, user and group ids are also shared, even if the associated user or group names are different between host and container)
- As such files that are created by processes running in the container by default will be owned by root, leading to permission issues when a process in the host wants to change/remove those files. As can be seen with [this failed build](https://buildkite.com/nubots/nubots/builds/5#b5178e6e-d916-4c45-be6b-0458051928d7/3-9), this lead to "permission denied" errors when cleaning out old build files for a new build.
- The [BuildKite recommendation](https://github.com/buildkite-plugins/docker-buildkite-plugin#propagate-uid-gid-optional-boolean) for this is to propagate the user/group id to the container, so files created in the container are owned by the corresponding host user. For this to work properly there needs to be a known user in the container to map to from the host.

There are [other ways](https://buildkite.com/docs/agent/v3/docker#permissions-errors-when-using-docker) to solve this problem, but this is one of the recommended  options.